File: Data Structures & Algorithms/interleaving-string/submission-3.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1)+len(s2) != len(s3): return False

        dp = {}

        def dfs(i, j):
            if i == len(s1) and j == len(s2):
                return True
            if (i, j) in dp: return dp[(i, j)]
            
            c1 = s1[i] if i < len(s1) else ""
            c2 = s2[j] if j < len(s2) else ""
            c3 = s3[i+j]

            if c1 == c2 and c1 == c3:
                dp[(i, j)] = dfs(i+1, j) or dfs(i, j+1)
            elif c1 == c3:
                dp[(i, j)] = dfs(i+1, j)
            elif c2 == c3:
                dp[(i, j)] = dfs(i, j+1)
            else:
                dp[(i, j)] = False

            return dp[(i, j)]

        return dfs(0, 0)

File: Data Structures & Algorithms/course-schedule-iv/submission-3.cpp
class Solution {
private:
    vector<vector<int>> adjList;
    unordered_map<int, unordered_set<int>> prereqMap;

    unordered_set<int>& dfs(int course) {
        if (prereqMap.contains(course))
            return prereqMap[course];

        for (auto& next : adjList[course]) {
            dfs(next);
            for (int prereq : prereqMap[next]) {
                prereqMap[course].insert(prereq);
            }
        }
        prereqMap[course].insert(course);

        return prereqMap[course];
    }

public:
    vector<bool> checkIfPrerequisite(int numCourses, vector<vector<int>>& prerequisites, vector<vector<int>>& queries) {
        adjList.resize(numCourses);
        for (auto& prereq : prerequisites) {
            adjList[prereq[0]].push_back(prereq[1]);
        }

        for (int i = 0; i < numCourses; ++i) {
            dfs(i);
        }

        vector<bool> ans(queries.size());
        for (int i = 0; i < queries.size(); ++i) {
            ans[i] = prereqMap[queries[i][0]].contains(queries[i][1]);
        }
        return ans;
    }
};